### PR TITLE
[builtins] Refactor cpu_model support to reduce #if nesting. NFCI

### PR DESCRIPTION
**Pull request by jroelofs**

### CODE REVIEW

File: compiler-rt/lib/builtins/cpu_model/aarch64.c [hunk 1]
```diff
@@ -0,0 +1,142 @@
+//===-- cpu_model_aarch64.c - Support for __cpu_model builtin  ----*- C -*-===//
```
Comment (jroelofs) [inline compiler-rt/lib/builtins/cpu_model/aarch64.c:1]: - [ ] FIXME: title comment has the wrong filename

File: compiler-rt/lib/builtins/cpu_model/aarch64.c [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+#  include "aarch64/fmv/mrs.inc"
+#  include "aarch64/fmv/android.inc"
+#elif __has_include(<sys/auxv.h>)
+#  include "aarch64/fmv/mrs.inc"
+#  include "aarch64/fmv/sysauxv.inc"
+#else
+#  include "aarch64/fmv/unimplemented.inc"
+#endif
+
+#endif // !defined(DISABLE_AARCH64_FMV)
```
Comment (jroelofs) [inline compiler-rt/lib/builtins/cpu_model/aarch64.c:142]: - [ ] FIXME: needs EOF newline

File: compiler-rt/lib/builtins/cpu_model/aarch64/lse_atomics/android.inc [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    // The kernel appears to have been corrected to mark it unsupported as of
+    // the Android 9.0 release on those devices, and this issue has not been
+    // observed anywhere else. Thus, this workaround may be removed if
+    // compiler-rt ever drops support for Android 8.0.
+    char arch[PROP_VALUE_MAX];
+    if (__isExynos9810())
+      result = false;
+  }
+  __aarch64_have_lse_atomics = result;
+}
```
Comment (jroelofs) [inline compiler-rt/lib/builtins/cpu_model/aarch64/lse_atomics/android.inc:28]: - [ ] FIXME: needs EOF newline

File: compiler-rt/lib/builtins/cpu_model/aarch64/lse_atomics/fucsia.inc [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+
+static void CONSTRUCTOR_ATTRIBUTE init_have_lse_atomics(void) {
+  // This ensures the vDSO is a direct link-time dependency of anything that
+  // needs this initializer code.
+#pragma comment(lib, "zircon")
+  uint32_t features;
+  zx_status_t status = _zx_system_get_features(ZX_FEATURE_KIND_CPU, &features);
+  __aarch64_have_lse_atomics =
+      status == ZX_OK && (features & ZX_ARM64_FEATURE_ISA_ATOMICS) != 0;
+}
```
Comment (jroelofs) [inline compiler-rt/lib/builtins/cpu_model/aarch64/lse_atomics/fucsia.inc:12]: - [ ] FIXME: needs EOF newline

File: compiler-rt/lib/builtins/cpu_model/aarch64/lse_atomics/sysauxv.inc [hunk 5]
```diff
@@ -0,0 +1,6 @@
+#include <sys/auxv.h>
+
+static void CONSTRUCTOR_ATTRIBUTE init_have_lse_atomics(void) {
+  unsigned long hwcap = getauxval(AT_HWCAP);
+  __aarch64_have_lse_atomics = (hwcap & HWCAP_ATOMICS) != 0;
+}
```
Comment (jroelofs) [inline compiler-rt/lib/builtins/cpu_model/aarch64/lse_atomics/sysauxv.inc:6]: - [ ] FIXME: needs EOF newline

File: compiler-rt/lib/builtins/cpu_model/cpu_model.h [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+// constructors in the end user application (starting at priority 101). This value
+// matches the libgcc choice for the same functions.
+#define CONSTRUCTOR_ATTRIBUTE __attribute__((constructor(90)))
+#else
+// FIXME: For MSVC, we should make a function pointer global in .CRT$X?? so that
+// this runs during initialization.
+#define CONSTRUCTOR_ATTRIBUTE
+#endif
+
+#endif
```
Comment (jroelofs) [inline compiler-rt/lib/builtins/cpu_model/cpu_model.h:41]: - [ ] FIXME: needs EOF newline

File: compiler-rt/lib/builtins/cpu_model/x86.c [hunk 7]
```diff
@@ -1,4 +1,4 @@
-//===-- cpu_model.c - Support for __cpu_model builtin  ------------*- C -*-===//
+//===-- cpu_model_x86.c - Support for __cpu_model builtin  --------*- C -*-===//
```
Comment (jroelofs) [inline compiler-rt/lib/builtins/cpu_model/x86.c:1]: - [ ] FIXME: rename file in title comment.

File: compiler-rt/lib/builtins/cpu_model/aarch64/fmv/android.inc [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+  // Android API 30.
+  int memfd_create(const char *, unsigned int) __attribute__((weak));
+  if (!memfd_create)
+    return;
+
+  __init_cpu_features_constructor(hwcap, arg);
+}
+
+void CONSTRUCTOR_ATTRIBUTE __init_cpu_features(void) {
+  unsigned long hwcap = 0;
```
Comment (jroelofs) [inline compiler-rt/lib/builtins/cpu_model/aarch64/fmv/android.inc:19]: - [ ] FIXME: dead variables

File: compiler-rt/lib/builtins/cpu_model/aarch64.c [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+// in Function Multi Versioning
+struct {
+  unsigned long long features;
+  // As features grows new fields could be added
+} __aarch64_cpu_features __attribute__((visibility("hidden"), nocommon));
+
+#if defined(__FreeBSD__)
+#include "aarch64/fmv/freebsd.inc"
+#include "aarch64/fmv/mrs.inc"
+#elif defined(__Fucsia__)
```
Comment (mstorsjo) [inline compiler-rt/lib/builtins/cpu_model/aarch64.c:127]: I believe you've misspelled Fuchsia consistently :-)

**Review by mstorsjo — COMMENTED** [review]
I think this looks reasonable overall, but there's a typo in the name of Fuchsia, which I believe would lead to its codepaths being entirely missed :-)

Thanks, in this form I think the ifdef jungle is much more manageable.

File: compiler-rt/lib/builtins/cpu_model/aarch64.c (hunk 9 above)
Comment (jroelofs) [reply compiler-rt/lib/builtins/cpu_model/aarch64.c:127]: 🤦‍♂️  thanks

**Review by ilinpv — COMMENTED** [review]
Thanks a lot for refactoring! It indeed becomes necessary and looks well-structured now, small nit - rename lse_atomics/fucsia.inc and fmv/fucsia.inc files as well.

**Comment by vitalybuka** [comment]
Breaks https://lab.llvm.org/buildbot/#/builders/269/builds/2975

**Comment by jroelofs** [comment]
There were a few follow-ups to address that. Sorry.

**Comment by jroelofs** [comment]
* c88e74c26d5c - (31 minutes ago) fixup! fixup! [builtins] Refactor cpu_model support to reduce #if nesting. NFCI - Jon Roelofs
* a5a17e8fadd4 - (39 minutes ago) fixup! [builtins] Refactor cpu_model support to reduce #if nesting. NFCI - Jon Roelofs
* 256b214b6b0f - (66 minutes ago) [builtins][arm64] Implement __init_cpu_features_resolver on Apple platforms (#75636) - Jon Roelofs
* 9237cfa65b6c - (68 minutes ago) [builtins] Refactor cpu_model support to reduce #if nesting. NFCI - Jon Roelofs
* b72e1609146e - (61 minutes ago) Revert "[builtins] Refactor cpu_model support to reduce #if nesting. NFCI" - Jon Roelofs
* b8b40e2fb988 - (61 minutes ago) Revert "[builtins][arm64] Implement __init_cpu_features_resolver on Apple platforms (#75636)" - Jon Roelofs
* 17aa52017103 - (66 minutes ago) [builtins][arm64] Implement __init_cpu_features_resolver on Apple platforms (#75636) - Jon Roelofs
* 025d048b1cac - (68 minutes ago) [builtins] Refactor cpu_model support to reduce #if nesting. NFCI - Jon Roelofs

**Comment by petrhosek** [comment]
We're still seeing the following error on our builders:
```
/b/s/w/ir/x/w/llvm_build/./bin/clang --target=aarch64-unknown-linux-gnu --sysroot=/b/s/w/ir/x/w/cipd/linux -DHAS_ASM_LSE -DVISIBILITY_HIDDEN  --target=aarch64-unknown-linux-gnu -O2 -g -DNDEBUG -fno-lto -std=c11 -fPIC -fno-builtin -fvisibility=hidden -fomit-frame-pointer -DCOMPILER_RT_HAS_FLOAT16 -MD -MT CMakeFiles/clang_rt.builtins-aarch64.dir/cpu_model/aarch64.c.o -MF CMakeFiles/clang_rt.builtins-aarch64.dir/cpu_model/aarch64.c.o.d -o CMakeFiles/clang_rt.builtins-aarch64.dir/cpu_model/aarch64.c.o -c /b/s/w/ir/x/w/llvm-llvm-project/compiler-rt/lib/builtins/cpu_model/aarch64.c
In file included from /b/s/w/ir/x/w/llvm-llvm-project/compiler-rt/lib/builtins/cpu_model/aarch64.c:43:
/b/s/w/ir/x/w/llvm-llvm-project/compiler-rt/lib/builtins/cpu_model/aarch64/lse_atomics/sysauxv.inc:5:41: error: use of undeclared identifier 'HWCAP_ATOMICS'
    5 |   __aarch64_have_lse_atomics = (hwcap & HWCAP_ATOMICS) != 0;
      |                                         ^
1 error generated.
```

**Comment by jroelofs** [comment]
@petrhosek oh, I see what the problem is. I'll add another follow-up shortly. Sorry for the noise.

**Comment by jroelofs** [comment]
More follow-ups in 85d5ed81b1df, and 52e7b6f5c520

**Comment by aeubanks** [comment]
still seeing Android errors
```
 [275/404] Building C object CMakeFiles/clang_rt.builtins-aarch64.dir/cpu_model/aarch64.c.o
 FAILED: CMakeFiles/clang_rt.builtins-aarch64.dir/cpu_model/aarch64.c.o 
 /b/s/w/ir/cache/builder/src/third_party/llvm-build/Release+Asserts/./bin/clang --target=aarch64-linux-android21 --sysroot=/b/s/w/ir/cache/builder/src/third_party/llvm-build-tools/debian_bullseye_amd64_sysroot -DHAS_ASM_LSE -DVISIBILITY_HIDDEN  --sysroot=/b/s/w/ir/cache/builder/src/third_party/android_toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/sysroot --unwindlib=none -mbranch-protection=standard -O2 -g -DNDEBUG -fno-lto -std=c11 -fPIC -fno-builtin -fvisibility=hidden -fomit-frame-pointer -DCOMPILER_RT_HAS_FLOAT16 -MD -MT CMakeFiles/clang_rt.builtins-aarch64.dir/cpu_model/aarch64.c.o -MF CMakeFiles/clang_rt.builtins-aarch64.dir/cpu_model/aarch64.c.o.d -o CMakeFiles/clang_rt.builtins-aarch64.dir/cpu_model/aarch64.c.o -c /b/s/w/ir/cache/builder/src/third_party/llvm/compiler-rt/lib/builtins/cpu_model/aarch64.c
 clang: warning: argument unused during compilation: '--unwindlib=none' [-Wunused-command-line-argument]
 In file included from /b/s/w/ir/cache/builder/src/third_party/llvm/compiler-rt/lib/builtins/cpu_model/aarch64.c:41:
 /b/s/w/ir/cache/builder/src/third_party/llvm/compiler-rt/lib/builtins/cpu_model/aarch64/lse_atomics/android.inc:6:3: error: void function '__isExynos9810' should not return a value [-Wreturn-type]
     6 |   return __system_property_get("ro.arch", arch) > 0 &&
       |   ^      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     7 |     strncmp(arch, "exynos9810", sizeof("exynos9810") - 1) == 0;
       |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 /b/s/w/ir/cache/builder/src/third_party/llvm/compiler-rt/lib/builtins/cpu_model/aarch64/lse_atomics/android.inc:11:25: error: call to undeclared function 'getauxval'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
    11 |   unsigned long hwcap = getauxval(AT_HWCAP);
       |                         ^
 /b/s/w/ir/cache/builder/src/third_party/llvm/compiler-rt/lib/builtins/cpu_model/aarch64/lse_atomics/android.inc:11:35: error: use of undeclared identifier 'AT_HWCAP'
    11 |   unsigned long hwcap = getauxval(AT_HWCAP);
       |                                   ^
 /b/s/w/ir/cache/builder/src/third_party/llvm/compiler-rt/lib/builtins/cpu_model/aarch64/lse_atomics/android.inc:12:27: error: use of undeclared identifier 'HWCAP_ATOMICS'
    12 |   _Bool result = (hwcap & HWCAP_ATOMICS) != 0;
       |                           ^
 /b/s/w/ir/cache/builder/src/third_party/llvm/compiler-rt/lib/builtins/cpu_model/aarch64/lse_atomics/android.inc:23:5: error: statement requires expression of scalar type ('void' invalid)
    23 |     if (__isExynos9810())
       |     ^   ~~~~~~~~~~~~~~~~
 In file included from /b/s/w/ir/cache/builder/src/third_party/llvm/compiler-rt/lib/builtins/cpu_model/aarch64.c:135:
 In file included from /b/s/w/ir/cache/builder/src/third_party/llvm/compiler-rt/lib/builtins/cpu_model/aarch64/fmv/mrs.inc:2:
 /b/s/w/ir/cache/builder/src/third_party/android_toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include/sys/auxv.h:53:19: error: conflicting types for 'getauxval'
    53 | unsigned long int getauxval(unsigned long int __type) __INTRODUCED_IN(18);
       |                   ^
 /b/s/w/ir/cache/builder/src/third_party/llvm/compiler-rt/lib/builtins/cpu_model/aarch64/lse_atomics/android.inc:11:25: note: previous implicit declaration is here
    11 |   unsigned long hwcap = getauxval(AT_HWCAP);
       |                         ^
 In file included from /b/s/w/ir/cache/builder/src/third_party/llvm/compiler-rt/lib/builtins/cpu_model/aarch64.c:136:
 /b/s/w/ir/cache/builder/src/third_party/llvm/compiler-rt/lib/builtins/cpu_model/aarch64/fmv/android.inc:25:3: error: statement requires expression of scalar type ('void' invalid)
    25 |   if (__isExynos9810())
       |   ^   ~~~~~~~~~~~~~~~~
 7 errors generated.
```

**Comment by jroelofs** [comment]
I am surprised I haven't seen any complaints from buildbots about Andoid... anyway, thanks.

394e481a38c774f12c765fde7e9302d039a7cd94

**Comment by petrhosek** [comment]
I have discovered another issue, on Fuchsia `libclang_rt.builtins.a` now has a reference to `getauxval` which it didn't have before. This is really undesirable because on Fuchsia, there's no `auxv` to begin with. I believe this is because before, the `__init_cpu_features` definition was guarded by:

```
#if defined(__has_include)
#if __has_include(<sys/auxv.h>)
#include <sys/auxv.h>
...
#if __has_include(<asm/hwcap.h>)
#include <asm/hwcap.h>
```
These `#if __has_include(...)` conditionals would evaluate to false on Fuchsia since we don't have those headers, but they have been moved in the refactored version.

**Comment by jroelofs** [comment]
@petrhosek I don't have a Fucshia sysroot to build this against, so I think it would help me a lot if you could grab the preprocessed version of the `cpu_model.c` before all these changes and stick it in a github gist... I feel a little bad for dragging this out in-tree.

> These #if __has_include(...) conditionals would evaluate to false on Fuchsia since we don't have those headers, but they have been moved in the refactored version.

In that case, if I'm reading it correctly, all of the `#if defined(__Fuchsia__)` support for both initializing the LSE atomics detection bool, and FMV in `cpu_model.c` are dead code, and we can drop those entries in the new `aarch64.c`.

**Pull request merged**